app/assets/javascripts/discourse/app/lib/notification-items/base.js

```diff
@@ -0,0 +1,114 @@
+import { formatUsername, postUrl } from "discourse/lib/utilities";
+import { userPath } from "discourse/lib/url";
+import { emojiUnescape } from "discourse/lib/text";
+import { htmlSafe } from "@ember/template";
+import I18n from "I18n";
+
+export default class NotificationItemBase {
+  constructor(notification, currentUser, siteSettings, site) {
+    this.notification = notification;
+    this.currentUser = currentUser;
+    this.siteSettings = siteSettings;
+    this.site = site;
+  }
+
+  /**
+   * @returns {string[]} An array of addtional classes that should be added to the <li> element of the notification item.
+   */
+  get classNames() {
+    return [];
+  }
+
+  /**
+   * @returns {string} A href/path that the notification item should link to.
+   */
+  get linkHref() {
+    if (this.topicId) {
+      return postUrl(
+        this.notification.slug,
+        this.topicId,
+        this.notification.post_number
+      );
+    }
+    if (this.notification.data.group_id) {
+      return userPath(
+        `${this.notification.data.username}/messages/${this.notification.data.group_name}`
+      );
+    }
+  }
+
+  /**
+   * @returns {string} A title for the notification item. It shows up when the user hovers over the notification item.
+   */
+  get linkTitle() {
+    if (this.notificationName) {
+      return I18n.t(`notifications.titles.${this.notificationName}`);
+    } else {
+      return "";
```

Comment:
I'm wondering if we should raise `not implemented` here since having no title is kind of bad for accessibility.

Reply:
It's risky because the only for the `else` branch to execute is if a notification type is no longer registered, but there are notification records with that type in the database (e.g. a plugin that adds a custom type is removed from a site). We attempt a best-effort rendering in situations like these, but raising a not implemented error would break the menu.

Reply:
I see, perhaps we can leave a comment here to indicate this possible case?

Reply:
done!